heidrun/hathi_marcxml.rb

```diff
@@ -0,0 +1,593 @@
+  },
+  'nbb' => {
+    'original_from' => 'Brooklyn Museum'
+  },
+  'njp' => {
+    'original_from' => 'Princeton University'
+  },
+  'nnc' => {
+    'original_from' => 'Columbia University'
+  },
+  'nnfr' => {
+    'original_from' => 'The Frick Collection'
+  },
+  'nrlf' => {
+    'original_from' => 'University of California'
+  },
+  'nwu' => {
+    'original_from' => 'Northwestern University'
+  },
+  'nyp' => {
+    'original_from' => 'New York Public Library'
+  },
+  'osu' => {
+    'original_from' => 'The Ohio State University'
+  },
+  'pst' => {
+    'original_from' => 'Pennsylvania State University'
+  },
+  'pur' => {
+    'original_from' => 'Purdue University'
+  },
+  'qmm' => {
+    'original_from' => 'McGill University'
+  },
+  'srlf' => {
+    'original_from' => 'University of California'
+  },
+  'txcm' => {
+    'original_from' => 'Texas A&amp;M University'
+  },
+  'ucbk' => {
+    'original_from' => 'University of California'
+  },
+  'ucd' => {
+    'original_from' => 'University of California'
+  },
+  'ucla' => {
+    'original_from' => 'University of California'
+  },
+  'ucm' => {
+    'original_from' => 'Universidad Complutense de Madrid'
+  },
+  'ucsc' => {
+    'original_from' => 'University of California'
+  },
+  'ucsd' => {
+    'original_from' => 'University of California'
+  },
+  'ucsf' => {
+    'original_from' => 'University of California'
+  },
+  'ufdc' => {
+    'original_from' => 'State University System of Florida'
+  },
+  'uiuc' => {
+    'original_from' => 'University of Illinois at Urbana-Champaign'
+  },
+  'uiucl' => {
+    'original_from' => 'University of Illinois at Urbana-Champaign'
+  },
+  'ukloku' => {
+    'original_from' => 'Knowledge Unlatched'
+  },
+  'umbus' => {
+    'original_from' => 'University of Michigan'
+  },
+  'umdb' => {
+    'original_from' => 'University of Michigan'
+  },
+  'umdcmp' => {
+    'original_from' => 'Millennium Project'
+  },
+  'umlaw' => {
+    'original_from' => 'University of Michigan'
+  },
+  'umn' => {
+    'original_from' => 'University of Minnesota'
+  },
+  'umprivate' => {
+    'original_from' => 'Private Donor'
+  },
+  'usu' => {
+    'original_from' => 'Utah State University Press'
+  },
+  'uuhhs' => {
+    'original_from' => 'Unitarian Universalist History and Heritage Society'
+  },
+  'uva' => {
+    'original_from' => 'University of Virginia'
+  },
+  'wau' => {
+    'original_from' => 'University of Washington'
+  },
+  'wu' => {
+    'original_from' => 'University of Wisconsin - Madison'
+  },
+  'yale' => {
+    'original_from' => 'Yale University'
+  }
+}
+
+data_provider_map = lambda do |r|
+  df_974 = Heidrun::MappingTools::MARC.datafield_els(r, '974')
+  df_974c = Heidrun::MappingTools::MARC.subfield_values(df_974, 'c')
+
+  df_974c.map do |raw_data_provider|
+    COLLECTION_CODES.fetch(raw_data_provider.downcase, nil)
+  end.compact.map { |collection| collection.fetch('original_from') }
```

Comment:
I was thrown a little by the structure of `COLLECTION_CODES`, where each element has `original_from`. It appears that `COLLECTION_CODES` is not used anywhere outside of this proc, so it's unclear to me why it's structured the way it is (as opposed to each value being a string, instead of a hash).  Is it because it is supposed to be copied and pasted conveniently from `ht_collections.json`?  Otherwise, the reason for the structure is not obvious. That's probably just a hint for an additional comment after line 208. 

If the `original_from` structure is necessary, It might be a little more clear also if the `.map` at line 425 were removed and line 424 were turned into ```COLLECTION_CODES.fetch(raw_data_provider.downcase, nil).fetch('original_from')```. I'm not sure if efficiency is an issue, but I think it would be faster.